libraries/AP_BattMonitor/AP_BattMonitor_TIBQ76952.cpp

```diff
@@ -0,0 +1,937 @@
+    int16_t cc2_current = direct_command_read_2bytes(TIBQ769x2_CC2Current);
+    accumulate.current += cc2_current * 0.001f; // convert to Amps
+
+    // read temperature
+    int16_t temp_internal = direct_command_read_2bytes(TIBQ769x2_IntTemperature); // 0.1K
+    accumulate.temp += KELVIN_TO_C(temp_internal * 0.1f); // convert to degC
+
+    // increment number of readings
+    accumulate.count++;
+}
+
+// ready battert power state (e.g. charging, discharging, idle)
+void AP_BattMonitor_TIBQ76952::read_power_state()
+{
+    AP_BattMonitor::PowerState new_state = _state.power_state;
+
+    uint16_t mfg_status = direct_command_read_2bytes(TIBQ769x2_MfgStatusInit);
+    uint8_t fet_status = direct_command_read_1byte(TIBQ769x2_FETStatus);
+
+    if (mfg_status & MfgStatusInit_FET_EN) {
+        new_state = AP_BattMonitor::PowerState::IDLE;
+        if (fet_status & FET_STATUS_CHG_FET_EN && accumulate.current > 0) {
+            // Charging if CHG_FET bit is set
+            new_state = AP_BattMonitor::PowerState::CHARGING;
+        }
+        if (fet_status & FET_STATUS_DSG_FET_EN) {
+            // Discharging if DSG_FET bit is set
+            new_state = AP_BattMonitor::PowerState::DISCHARGING;
+        }
+    }
+
+    _state.power_state = new_state;
+}
+
+// ready safety status (e.g. check for faults)
+void AP_BattMonitor_TIBQ76952::read_safety_state()
+{
+    uint16_t safetyA_status = direct_command_read_2bytes(TIBQ769x2_SafetyStatusA);
+    if (safetyA_status > 0) {
+        Debug("BQ76952: SafetyA Status: 0x%04X", safetyA_status);
+    }
+}
+
+// read bytes from a register. returns true on success
+bool AP_BattMonitor_TIBQ76952::read_register(uint8_t reg_addr, uint8_t *reg_data, uint8_t len) const
+{
+    WITH_SEMAPHORE(dev->get_semaphore());
+
+    // read bytes from device
+    return dev->read_registers(reg_addr, reg_data, len);
+}
+
+// write a single byte to consecutive registers. returns true on success
+bool AP_BattMonitor_TIBQ76952::write_register(uint8_t reg_addr, const uint8_t *reg_data, uint8_t len) const
+{
+    WITH_SEMAPHORE(dev->get_semaphore());
+
+    // write each byte
+    for (uint8_t i = 0; i < len; i++) {
+        if (!dev->write_register(reg_addr + i, reg_data[i])) {
+            return false;
+        }
+    }
+    return true;
+}
+
+// read or write a direct command. returns true on success
+// direct commands are send using a 7-bit command address and may trigger an action or read/write a datavalue
+bool AP_BattMonitor_TIBQ76952::direct_command(uint16_t command, uint16_t data, CommandType type, uint8_t *rx_data, uint8_t rx_len) const
+{
+    // sanity check read buffer
+    if (type == CommandType::READ && rx_data == nullptr) {
+        return false;
+    }
+
+    // prepare transmit buffer
+    uint8_t tx_buffer[2] = {LOWBYTE(data), HIGHBYTE(data)};
+    switch (type) {
+    case CommandType::READ:
+        read_register(command, rx_data, rx_len);
+        break;
```

Comment:
```suggestion
        return read_register(command, rx_data, rx_len);
```

Reply:
txs, we now directly return the result of the read_register() call